### PR TITLE
#314 Treat HTTP requests containing a JSON accept header as a JSON request

### PR DESCRIPTION
**Pull request by elight**
For the purposes of deciding if and how to treat the body of a request, respect any JSON type specified in the Accept header.

### CODE REVIEW

**Comment by elight** [comment]
Addresses #314. Would love to see this merged and released soon. Thanks!

File: lib/rollbar/request_data_extractor.rb [hunk 1]
```diff
@@ -130,6 +135,11 @@ def rollbar_raw_body_params(rack_req)
       rack_req.body.rewind
     end
 
+    def request_has_json_body?(rack_req)
+      rack_req.env['CONTENT_TYPE'] =~ %r{application/json} ||
+      rack_req.env['ACCEPT']       =~ /\bjson\b/
```
Comment (jingweno) [inline lib/rollbar/request_data_extractor.rb]: This regexp may not work for a typical `ACCEPT` header: `Accept: application/vnd.heroku+json; version=3`

Comment (jingweno) [inline lib/rollbar/request_data_extractor.rb]: Never mind. `\b` matches word boundary.

**Comment by dmcinnes** [comment]
Nice :+1:

File: lib/rollbar/request_data_extractor.rb [hunk 2]
```diff
@@ -28,7 +28,12 @@ def extract_request_data_from_rack(env)
       session = rollbar_filtered_params(sensitive_params, rollbar_request_session(rack_req))
       route_params = rollbar_filtered_params(sensitive_params, rollbar_route_params(env))
 
-      params = request_params.merge(get_params).merge(post_params).merge(raw_body_params)
+      params =
+        if !raw_body_params.empty?
```
Comment (jondeandres) [inline lib/rollbar/request_data_extractor.rb]: @elight this change is not related to the PR title. We may probably discuss this in another one.

**Comment by jondeandres** [comment]
Thanks @elight for the PR :smile:. I left a comment https://github.com/rollbar/rollbar-gem/pull/322#discussion_r42397639, apart of that this is ok for me.

File: lib/rollbar/request_data_extractor.rb (hunk 2 above)
Comment (elight) [inline lib/rollbar/request_data_extractor.rb]: After making my changes, **without** L31-36 above, I eventually found that `params` was getting the post params twice: once as the desired hash and once in some serialized form attached to a key called `key`, IIRC. This seems to be because the `post_params` contained this undesirable value while the `raw_body_params` had what I needed to get the new tests passing while not breaking the old tests.

File: lib/rollbar/request_data_extractor.rb (hunk 1 above)
Comment (elight) [inline lib/rollbar/request_data_extractor.rb]: `\b` is a Good Thing™ ;-)

File: lib/rollbar/request_data_extractor.rb (hunk 2 above)
Comment (jondeandres) [inline lib/rollbar/request_data_extractor.rb]: @elight I'd say we are getting the `post_params` and `raw_body_params` cause Rack doesn't parse JSON payload when you ask for `request.POST`. But you are probably right in the sense that we are doing something wrong. We'll look in detail to this.

But anyway, I think this logic shouldn't be here in this PR. What do you think?

Comment (elight) [inline lib/rollbar/request_data_extractor.rb]: @jondeandres The PR doesn't pass specs without this logic.

Comment (jondeandres) [inline lib/rollbar/request_data_extractor.rb]: Hi @elight we cannot change that logic in the way you changed it. You are not adding the GET params to the `params` variable.

For `request_params` and `post_params` we may need to change some logic. But this cannot be merged with this change. Probably tests should be changed to fit the new behavior, but this implementation cannot be released with this change.

If you want to investigate what's happening in those you are welcome, if you don't a free slot to do that just notice us and we'll investigate it.

Thanks!

**Comment by elight** [comment]
@jondeandres Seems that the exclusion of `CONTENT_TYPE` causes the `post_params` to come through as `{"{\"key\":\"value\"}"=>nil}`. Updated accordingly to handle when the `CONTENT_TYPE` is absent and `ACCEPT` is a JSON type. Sorry for the awkward back and forth before!

File: lib/rollbar/request_data_extractor.rb [hunk 3]
```diff
@@ -130,6 +136,21 @@ def rollbar_raw_body_params(rack_req)
       rack_req.body.rewind
     end
 
+    def json_request?(rack_req)
+      rack_req.env['CONTENT_TYPE'] =~ %r{application/json} ||
+      rack_req.env['ACCEPT']       =~ /\bjson\b/
+    end
+
+    def json_request_missing_content_type?(env)
```
Comment (elight) [inline lib/rollbar/request_data_extractor.rb]: This is a little clumsier than I would like–but the added work on L1456-147 is necessary to abstract this method.

**Comment by elight** [comment]
Looks like the tests still pass if I merge the `raw_body_params` into the `get_params`. Is this adequate?

As @jingweno pointed out, it's legal for JSON requests to include an `Accept` and not a `ContentType`. As far as I can tell, this change is necessary to support that.

**Comment by jondeandres** [comment]
@elight thanks, let me play with this branch and I'll notify about it ok?

Thanks again!

**Comment by elight** [comment]
Sure thing!
